# Customer Churn Prediction Project
## Task 1: Data Importation and Cleaning | Task 2: Data Wrangling and Manipulation



---
# Task 1: Data Importation and Cleaning

In [1]:
import pandas as pd
import numpy as np

pd.set_option("display.max_columns", None)

## 1. Import the dataset

In [2]:
df = pd.read_csv(r"C:\Users\mfana\CDAM_09-26_IP028_CustomerChurnPrediction_Kinyua\data\Customer_churn_detection_data.csv")
print(f"Dataset loaded: {df.shape[0]} rows, {df.shape[1]} columns")

Dataset loaded: 10000 rows, 20 columns


## 2. Display first and last observations

In [3]:
df.head()

,Customer_ID,Age,Gender,Marital_Status,Education_Level,Employment_Status,Account_Tenure_Months,Account_Type,Number_of_Products,Average_Balance,Monthly_Transactions,Online_Banking_Usage,Mobile_Banking_Usage,Number_of_Complaints,Complaint_Resolution_Rate,Customer_Support_Calls,Email_Opt_In,Credit_Card_Holder,Loan_Account_Holder,Churn
0,CUST2463,32,Male,Divorced,Primary,Employed,13,Current,2,57001,19,6,10,0,0.97,0,1,0,0,1
1,CUST2511,45,Male,Single,University,Employed,88,Current,1,24533,14,5,9,0,0.89,2,1,0,0,0
2,CUST2227,29,Female,Single,Primary,Employed,151,Savings,3,212133,16,11,7,1,0.71,2,1,0,0,0
3,CUST0526,49,Female,Single,Secondary,Employed,118,Savings,2,94045,22,12,8,0,0.89,0,1,0,0,0
4,CUST0195,33,Female,Single,Primary,Employed,108,Savings,1,137693,15,7,8,0,0.64,1,0,0,0,0


In [4]:
df.tail()

,Customer_ID,Age,Gender,Marital_Status,Education_Level,Employment_Status,Account_Tenure_Months,Account_Type,Number_of_Products,Average_Balance,Monthly_Transactions,Online_Banking_Usage,Mobile_Banking_Usage,Number_of_Complaints,Complaint_Resolution_Rate,Customer_Support_Calls,Email_Opt_In,Credit_Card_Holder,Loan_Account_Holder,Churn
9995,CUST1372,32,Female,Single,Primary,Employed,103,Savings,1,65036,18,11,8,1,0.87,1,1,1,0,0
9996,CUST2242,18,Male,Married,Secondary,Employed,143,Current,3,12053,17,7,11,1,0.57,0,1,1,0,0
9997,CUST0057,47,Male,Single,Primary,Self-employed,164,Current,1,260336,19,8,7,0,0.77,1,0,0,0,0
9998,CUST1354,27,Male,Single,Secondary,Employed,146,Current,3,63896,17,7,6,0,0.69,0,1,0,0,0
9999,CUST2681,27,Female,Single,Primary,Employed,175,Savings,1,102781,12,6,6,0,0.93,0,1,0,1,0


## 3. Variable names and data types

In [5]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 20 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Customer_ID                10000 non-null  str    
 1   Age                        10000 non-null  int64  
 2   Gender                     10000 non-null  str    
 3   Marital_Status             10000 non-null  str    
 4   Education_Level            8974 non-null   str    
 5   Employment_Status          10000 non-null  str    
 6   Account_Tenure_Months      10000 non-null  int64  
 7   Account_Type               10000 non-null  str    
 8   Number_of_Products         10000 non-null  int64  
 9   Average_Balance            10000 non-null  int64  
 10  Monthly_Transactions       10000 non-null  int64  
 11  Online_Banking_Usage       10000 non-null  int64  
 12  Mobile_Banking_Usage       10000 non-null  int64  
 13  Number_of_Complaints       10000 non-null  int64  
 14  Co

## 4. Dataset dimensions

In [6]:
print(f"Rows: {df.shape[0]}")
print(f"Columns: {df.shape[1]}")

Rows: 10000
Columns: 20


## 5. Descriptive statistics

In [7]:
df.describe()

,Age,Account_Tenure_Months,Number_of_Products,Average_Balance,Monthly_Transactions,Online_Banking_Usage,Mobile_Banking_Usage,Number_of_Complaints,Complaint_Resolution_Rate,Customer_Support_Calls,Email_Opt_In,Credit_Card_Holder,Loan_Account_Holder,Churn
count,10000.00000,10000.000000,10000.000000,1.000000e+04,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000,10000.00000,10000.000000,10000.000000,10000.000000,10000.000000
mean,35.15840,90.817100,2.005800,1.501083e+05,15.053100,8.00500,10.027600,0.193300,0.747475,1.01120,0.691600,0.298300,0.206900,0.273200
std,9.62961,51.771835,1.005021,1.513211e+05,3.870294,2.85198,3.175791,0.437898,0.144261,1.00746,0.461856,0.457535,0.405103,0.445625
min,18.00000,1.000000,1.000000,5.000000e+02,3.000000,0.00000,0.000000,0.000000,0.500000,0.00000,0.000000,0.000000,0.000000,0.000000
25%,28.00000,46.000000,1.000000,4.422200e+04,12.000000,6.00000,8.000000,0.000000,0.620000,0.00000,0.000000,0.000000,0.000000,0.000000
50%,35.00000,91.000000,2.000000,1.036505e+05,15.000000,8.00000,10.000000,0.000000,0.750000,1.00000,1.000000,0.000000,0.000000,0.000000
75%,42.00000,136.000000,3.000000,2.065610e+05,18.000000,10.00000,12.000000,0.000000,0.870000,2.00000,1.000000,1.000000,0.000000,1.000000
max,70.00000,180.000000,4.000000,1.644080e+06,32.000000,21.00000,27.000000,3.000000,1.000000,7.00000,1.000000,1.000000,1.000000,1.000000


In [8]:
df.describe(include="object")

C:\Users\mfana\AppData\Local\Temp\ipykernel_33704\702825166.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include="object")


,Customer_ID,Gender,Marital_Status,Education_Level,Employment_Status,Account_Type
count,10000,10000,10000,8974,10000,10000
unique,2899,2,3,3,3,3
top,CUST1448,Female,Married,Secondary,Employed,Savings
freq,13,5035,5036,3995,6055,5932


## 6. Check for duplicate observations
Two checks: (a) fully identical rows, (b) duplicate `Customer_ID` — a repeated ID with
different data is a different kind of problem than a true duplicate row.

In [9]:
full_dupes = df.duplicated().sum()
print(f"Fully duplicate rows: {full_dupes}")

id_dupes = df["Customer_ID"].duplicated().sum()
print(f"Rows with a repeated Customer_ID: {id_dupes} out of {df.shape[0]}")
print(f"Unique Customer_IDs: {df['Customer_ID'].nunique()}")

Fully duplicate rows: 0
Rows with a repeated Customer_ID: 7101 out of 10000
Unique Customer_IDs: 2899


**Finding:** 0 fully duplicate rows, but 7,101 rows share a `Customer_ID` with at least
one other row — and those rows have different Age, Gender, and account details. `Customer_ID`
was generated from a smaller pool of values and reused; it is **not** a reliable unique
customer key in this dataset. Each row is still treated as an independent customer record.
This is a data quality limitation worth stating explicitly in the report, not something we
silently "fix".

## 7. Identify missing values

In [10]:
missing = df.isnull().sum()


**Finding:** `Education_Level` has 1,026 missing values (10.26%) — the only column affected.

## 8. Handle missing values
For a categorical column with ~10% missing and no obvious relationship to force a specific
value, an explicit **"Unknown"** category is safer than imputing the mode — imputing would
artificially inflate one education group and could bias the model. This also preserves the
missingness itself as potential signal.

In [11]:
df["Education_Level"] = df["Education_Level"].fillna("Unknown")
df["Education_Level"].value_counts()

Education_Level
Secondary     3995
Primary       2961
University    2018
Unknown       1026
Name: count, dtype: int64

## 9. Check for impossible or unrealistic observations

In [12]:
checks = {
    "Negative Age": (df["Age"] < 0).sum(),
    "Age unrealistic (<18 or >100)": ((df["Age"] < 18) | (df["Age"] > 100)).sum(),
    "Negative Account_Tenure_Months": (df["Account_Tenure_Months"] < 0).sum(),
    "Negative Average_Balance": (df["Average_Balance"] < 0).sum(),
    "Negative Monthly_Transactions": (df["Monthly_Transactions"] < 0).sum(),
    "Complaint_Resolution_Rate out of [0,1]": (
        (df["Complaint_Resolution_Rate"] < 0) | (df["Complaint_Resolution_Rate"] > 1)
    ).sum(),
    "Number_of_Products <= 0": (df["Number_of_Products"] <= 0).sum(),
    "Churn not in {0,1}": (~df["Churn"].isin([0, 1])).sum(),
}
for check, count in checks.items():
    print(f"{check}: {count}")

Negative Age: 0
Age unrealistic (<18 or >100): 0
Negative Account_Tenure_Months: 0
Negative Average_Balance: 0
Negative Monthly_Transactions: 0
Complaint_Resolution_Rate out of [0,1]: 0
Number_of_Products <= 0: 0
Churn not in {0,1}: 0


**Finding:** No impossible values found — all numeric fields fall within realistic
bounds (Age 18–70, non-negative balances, resolution rate in [0,1], Churn strictly 0/1).

## 10. Save the cleaned dataset

In [13]:
df.to_csv("../data/churn_cleaned.csv", index=False)
print(f"Final shape: {df.shape}")

Final shape: (10000, 20)


### Summary of cleaning steps 
1. Loaded 10,000 rows x 20 columns.
2. Checked for fully duplicate rows — none found.
3. Found `Customer_ID` repeats across unrelated records (2,899 unique IDs across 10,000 rows)
   — flagged as a data quality limitation, not treated as duplication.
4. Found 1,026 missing values (10.26%) in `Education_Level` only.
5. Imputed missing `Education_Level` with an explicit "Unknown" category.
6. Ran realism checks on Age, tenure, balance, transactions, complaint resolution rate,
   product count, and the churn label — no impossible values found.
7. Saved the cleaned dataset to `data/cleaned/churn_cleaned.csv`.

---
# Task 2: Data Wrangling and Manipulation

## 1. Select specific variables
customer identifier plus the demographic, account, and behavioural variables are 
most relevant to a churn analysis.

In [14]:
selected = df[[
    "Customer_ID", "Age", "Gender", "Employment_Status", "Account_Type",
    "Account_Tenure_Months", "Number_of_Products", "Average_Balance",
    "Monthly_Transactions", "Online_Banking_Usage", "Mobile_Banking_Usage",
    "Number_of_Complaints", "Complaint_Resolution_Rate", "Churn"
]].copy()
selected.head(10)

,Customer_ID,Age,Gender,Employment_Status,Account_Type,Account_Tenure_Months,Number_of_Products,Average_Balance,Monthly_Transactions,Online_Banking_Usage,Mobile_Banking_Usage,Number_of_Complaints,Complaint_Resolution_Rate,Churn
0,CUST2463,32,Male,Employed,Current,13,2,57001,19,6,10,0,0.97,1
1,CUST2511,45,Male,Employed,Current,88,1,24533,14,5,9,0,0.89,0
2,CUST2227,29,Female,Employed,Savings,151,3,212133,16,11,7,1,0.71,0
3,CUST0526,49,Female,Employed,Savings,118,2,94045,22,12,8,0,0.89,0
4,CUST0195,33,Female,Employed,Savings,108,1,137693,15,7,8,0,0.64,0
5,CUST2986,33,Female,Self-employed,Savings,69,3,63153,15,9,11,0,0.78,1
6,CUST1842,41,Male,Employed,Current,167,2,378881,11,7,14,1,0.63,0
7,CUST1142,41,Female,Self-employed,Current,68,1,43330,10,10,7,0,0.95,0
8,CUST1253,49,Female,Self-employed,Current,78,4,307651,18,7,9,1,0.87,1
9,CUST1268,40,Female,Employed,Savings,143,2,105577,16,9,8,0,0.66,0


## 2. Filter observations based on conditions
Restricting to **established customers** (tenure over 12 months) since the new customers 
 can distort early patterns.

In [15]:
filtered = selected[selected["Account_Tenure_Months"] > 12].copy()
print(f"Rows before filter: {selected.shape[0]}")
print(f"Rows after filter (tenure > 12 months): {filtered.shape[0]}")

Rows before filter: 10000
Rows after filter (tenure > 12 months): 9343


## 3. Create at least two new variables
- `Total_Digital_Usage`: combined online + mobile banking usage for a single unit of easure
- `Balance_per_Product`: average balance spread across the number of products held which helps explain  how "deep" the relationship is per product.

In [16]:
filtered["Total_Digital_Usage"] = filtered["Online_Banking_Usage"] + filtered["Mobile_Banking_Usage"]
filtered["Balance_per_Product"] = (filtered["Average_Balance"] / filtered["Number_of_Products"]).round(2)
filtered[["Total_Digital_Usage", "Balance_per_Product"]].describe()

,Total_Digital_Usage,Balance_per_Product
count,9343.000000,9.343000e+03
mean,18.029969,9.602141e+04
std,4.255962,1.165975e+05
min,4.000000,1.250000e+02
25%,15.000000,2.288800e+04
50%,18.000000,5.685300e+04
75%,21.000000,1.234705e+05
max,36.000000,1.230074e+06


## 4. Rename variables where appropriate
Shortening names for clarity in later tables and plots.

In [17]:
filtered = filtered.rename(columns={
    "Account_Tenure_Months": "Tenure_Months",
    "Average_Balance": "Avg_Balance"
})
filtered.columns.tolist()

['Customer_ID',
 'Age',
 'Gender',
 'Employment_Status',
 'Account_Type',
 'Tenure_Months',
 'Number_of_Products',
 'Avg_Balance',
 'Monthly_Transactions',
 'Online_Banking_Usage',
 'Mobile_Banking_Usage',
 'Number_of_Complaints',
 'Complaint_Resolution_Rate',
 'Churn',
 'Total_Digital_Usage',
 'Balance_per_Product']

## 5. Sort observations
Sorting by churn status, then by balance - this will show the highest-value churned customers first.

In [18]:
sorted_df = filtered.sort_values(by=["Churn", "Avg_Balance"], ascending=[False, False])
sorted_df.head(10)

,Customer_ID,Age,Gender,Employment_Status,Account_Type,Tenure_Months,Number_of_Products,Avg_Balance,Monthly_Transactions,Online_Banking_Usage,Mobile_Banking_Usage,Number_of_Complaints,Complaint_Resolution_Rate,Churn,Total_Digital_Usage,Balance_per_Product
3843,CUST2690,29,Male,Employed,Savings,168,2,1644080,13,10,7,0,0.51,1,17,822040.00
9628,CUST0912,37,Female,Unemployed,Savings,56,4,1442429,13,8,7,2,0.90,1,15,360607.25
1862,CUST0678,35,Male,Self-employed,Savings,104,1,1060248,14,10,11,1,0.60,1,21,1060248.00
2209,CUST2913,22,Male,Employed,Fixed Deposit,33,4,1059229,16,5,10,0,0.95,1,15,264807.25
2390,CUST2580,47,Female,Self-employed,Fixed Deposit,151,1,1006257,19,13,12,0,0.50,1,25,1006257.00
1602,CUST2668,21,Male,Self-employed,Current,40,3,954288,10,3,7,0,0.69,1,10,318096.00
7650,CUST2970,25,Female,Employed,Savings,103,2,936368,16,8,10,0,0.88,1,18,468184.00
2339,CUST1973,50,Male,Self-employed,Savings,37,1,899628,13,9,14,0,0.69,1,23,899628.00
1263,CUST2455,28,Male,Employed,Current,127,1,890509,15,8,13,1,0.70,1,21,890509.00
909,CUST0833,36,Male,Self-employed,Savings,76,1,860216,15,11,11,0,0.51,1,22,860216.00


## 6 & 7. Group by a categorical variable and compute summary statistics

In [19]:
summary_table = filtered.groupby("Account_Type").agg(
    Avg_Balance=("Avg_Balance", "mean"),
    Avg_Tenure_Months=("Tenure_Months", "mean"),
    Avg_Digital_Usage=("Total_Digital_Usage", "mean"),
    Churn_Rate=("Churn", "mean"),
    Customer_Count=("Customer_ID", "count")
).round(2).sort_values("Churn_Rate", ascending=False)
summary_table

,Avg_Balance,Avg_Tenure_Months,Avg_Digital_Usage,Churn_Rate,Customer_Count
Account_Type,,,,,
Savings,150376.96,97.09,18.06,0.27,5569
Current,149281.85,95.66,18.05,0.26,2823
Fixed Deposit,152345.48,97.84,17.79,0.25,951


## 8. Clean summary table
Same table, formatted for direct use in the report (percentages and currency-style formatting).

In [20]:
report_table = summary_table.copy()
report_table["Churn_Rate"] = (report_table["Churn_Rate"] * 100).round(1).astype(str) + "%"
report_table["Avg_Balance"] = report_table["Avg_Balance"].apply(lambda x: f"{x:,.0f}")
report_table

,Avg_Balance,Avg_Tenure_Months,Avg_Digital_Usage,Churn_Rate,Customer_Count
Account_Type,,,,,
Savings,"150,377",97.09,18.06,27.0%,5569
Current,"149,282",95.66,18.05,26.0%,2823
Fixed Deposit,"152,345",97.84,17.79,25.0%,951


###  interpretation
- Filtering to established customers (tenure > 12 months) dropped the dataset from 10,000 to
  9,343 rows — most customers are past the one-year mark, so this filter mainly removes very
  new accounts rather than a large chunk of the base.
- **Account type shows only a small difference in churn rate** — Savings accounts churn
  slightly more (≈27%) than Fixed Deposit accounts (≈25%), but the gap is narrow. This
  suggests account type alone is a weak churn predictor on its own, and the model in Task 4
  will need to combine it with behavioural variables (complaints, digital usage, balance) to
  find stronger signal — a useful, honest finding to report rather than overselling it.
- Fixed Deposit customers carry the highest average balance and tenure, consistent with that
  product being a longer-term commitment.
- `Total_Digital_Usage` and `Balance_per_Product` are now available as engineered features for
  the EDA (Task 3) and modelling (Task 4) stages.

## Save the wrangled dataset

In [21]:
filtered.to_csv("../data/churn_wrangled.csv", index=False)
print("Wrangled dataset saved to data/churn_wrangled.csv")
print(f"Final shape: {filtered.shape}")

Wrangled dataset saved to data/churn_wrangled.csv
Final shape: (9343, 16)
